pkg/node/local_node_store.go

```diff
@@ -107,40 +99,28 @@ func NewLocalNodeStore(params LocalNodeStoreParams) (LocalNodeStore, error) {
 			s.complete = nil
 			s.emit = nil
 			s.mu.Unlock()
+
+			localNode = nil
 			return nil
 		},
 	})
 
 	return s, nil
 }
 
-// defaultLocalNodeStore constructs the default instance for the LocalNodeStore used by
-// address.go.
-func defaultLocalNodeStore() LocalNodeStore {
-	src, emit, complete := stream.Multicast[LocalNode](stream.EmitLatest)
-	s := &localNodeStore{
-		Observable: src,
-		valid:      true,
-		emit:       emit,
-		complete:   complete,
-	}
-	s.cond = sync.NewCond(&s.mu)
-	return s
-}
-
-func (s *localNodeStore) Get() LocalNode {
+// Get retrieves the current local node. Use Get() only for inspecting the state,
+// e.g. in API handlers. Do not assume the value does not change over time.
+// Blocks until the store has been initialized.
```

Comment:
^ Side note while trying to answer my own questions here: I got confused for a while searching for the implementation of `Observe()` to satisfy the `Observable` interface. It would help to expand the comment above the `Observable[]` interface in `pkg/stream/observable.go` to explain the common usage by embedding this field, and that the implementation of the `Observe()` function is expected to be provided by other types in pkg/stream, provided by `FuncObservable`. This could potentially help others to explore this code in future and understand the implementation if they intend to dig through and confirm properties like the blocking property above.

Reply:
Maybe coming back to this, it seems like maybe the intent behind the "Blocking" property is that the `s.mu` will stall while trying to grab the mutex, but it doesn't look like initialization grabs the mutex, so initial calls to `Get()` might not block (?)

Reply:
Taking the lock in `Get` is wrong. If someone calls `Get` in the background before `LocalNodeStore` has started it'll block its start hook. While it's not expected that `LocalNodeStore` is used before it has been started and initialized, I'll still try to make this safe to do so.

Reply:
Let's say that someone manages to get a `LocalNodeStore` pointer and calls `Get()`. Where will the block occur exactly? Presumably that `Get()` call descends into `stream.First()` -> `s.Observe()` -> (???).

Alternatively someone can only get a `LocalNodeStore` pointer via Hive after the start hook has run, in which case this comment doesn't really apply to this specific function?

Reply:
If you get the `LocalNodeStore` in your constructor and then fork a goroutine that calls `Get` it'll block until `LocalNodeStore` start hook runs and it gets initialized and the first `emit` call is made.

Reply:
OK, so tracing through:
- Some code calls `Get()`
- Get() calls `First()`
- First does some non-blocking operations and then calls `src.Observe()`
- `Observe()` is implemented by the anonymous function in `Multicast()`, which may grab function-level locks (should not be held) and spins out goroutines, then returns
- `First()` provides a `next()` function that grabs the first item and cancels the context immediately
- When `First()` returns from `Observe()`, it blocks on the completion error channel, which is closed as soon as the first element is emitted (per the function above).
- The anonymous goroutines from the `Observe()` call will only close the completion / err channel once the first element is emitted
- The first element is only emitted by the code calling `Multicast()`'s returned `emit()` function, which is only triggered by an `OnStart()` cell lifecycle hook.

From that, I conclude that `pkg/stream:First()` function should have in its API description "This function blocks until the first element is emitted into the stream".

That makes me wonder about the `io.EOF` case in `First()` though, how can that happen if it blocks until the first element is emitted? :thinking:

Reply:
https://github.com/cilium/cilium/pull/25922 -- doc improvements here